File: appmap/links.py
"""Link derivation — the one bit of 'graph' we keep this pass.

Navigability comes from surface->surface links, so we compute exactly what
strengthens that:
  * incoming-edge inversion (backlinks per surface), and
  * dangling-link detection (an edge/entry_point/containment target that names a
    surface id that doesn't exist).

No reachability/orphan/dead-end analysis here — that's deferred.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .model import Surface


@dataclass
class Incoming:
    """A derived backlink: `from_id` reaches this surface via `edge`."""

    from_id: str
    edge: dict[str, Any]


@dataclass
class DanglingLink:
    """A reference to a surface id that has no record."""

    from_id: str
    to: str
    kind: str  # "edge" | "entry_point" | "contains"
    ref: str   # edge id / entry-point key / containment target, for locating it


@dataclass
class LinkGraph:
    incoming: dict[str, list[Incoming]] = field(default_factory=dict)
    dangling: list[DanglingLink] = field(default_factory=list)


def build_links(surfaces: list[Surface]) -> LinkGraph:
    ids = {s.id for s in surfaces}
    incoming: dict[str, list[Incoming]] = {s.id: [] for s in surfaces}
    dangling: list[DanglingLink] = []

    for s in surfaces:
        for edge in s.edges:
            to = edge.get("to")
            if not to:
                continue
            if to in ids:
                incoming[to].append(Incoming(from_id=s.id, edge=edge))
            else:
                dangling.append(
                    DanglingLink(from_id=s.id, to=to, kind="edge", ref=edge.get("id", to))
                )

        for child in s.contains:
            if child not in ids:
                dangling.append(
                    DanglingLink(from_id=s.id, to=child, kind="contains", ref=child)
                )

        for ep in s.entry_points:
            to = ep.get("to")
            if to and to not in ids:
                key = f"{ep.get('type', '?')}:{ep.get('value', '')}"
                dangling.append(
                    DanglingLink(from_id=s.id, to=to, kind="entry_point", ref=key)
                )

    return LinkGraph(incoming=incoming, dangling=dangling)

File: cli/Sources/AppMapKit/Links.swift
// Link derivation: incoming-edge inversion (backlinks) and dangling-link
// detection. Incoming is never stored in records — always derived here.

import Foundation

public struct Incoming {
    public let fromID: String
    public let edge: [String: Any]
}

public struct DanglingLink {
    public let fromID: String
    public let to: String
    public let kind: String  // "edge" | "entry_point" | "contains"
    public let ref: String   // edge id / entry-point key / containment target
}

public struct LinkGraph {
    public var incoming: [String: [Incoming]] = [:]
    public var dangling: [DanglingLink] = []
}

public func buildLinks(_ surfaces: [SurfaceRecord]) -> LinkGraph {
    let ids = Set(surfaces.map(\.id))
    var graph = LinkGraph()
    for s in surfaces { graph.incoming[s.id] = [] }

    for s in surfaces {
        for edge in s.edges {
            guard let to = edge["to"] as? String, !to.isEmpty else { continue }
            if ids.contains(to) {
                graph.incoming[to, default: []].append(Incoming(fromID: s.id, edge: edge))
            } else {
                graph.dangling.append(DanglingLink(
                    fromID: s.id, to: to, kind: "edge",
                    ref: edge["id"] as? String ?? to
                ))
            }
        }
        for child in s.contains where !ids.contains(child) {
            graph.dangling.append(DanglingLink(
                fromID: s.id, to: child, kind: "contains", ref: child
            ))
        }
        for ep in s.entryPoints {
            guard let to = ep["to"] as? String, !to.isEmpty, !ids.contains(to) else { continue }
            let key = "\(ep["type"] as? String ?? "?"):\(ep["value"] as? String ?? "")"
            graph.dangling.append(DanglingLink(
                fromID: s.id, to: to, kind: "entry_point", ref: key
            ))
        }
    }
    return graph
}
